### ISSUE TITLE
Getting degrading results from Sam2 compared to Sam Segmentations

### ISSUE THREAD

**Issue by omrastogi** [facebookresearch/segment-anything-2#93]
## SAM

- Version: `vit_l `
- Input type: `box input` 
- Multimask Output: `True`

![viz_7_sam](https://github.com/user-attachments/assets/ccfe66c8-938f-470a-9c7a-819839cc6ee7)
![viz_13_sam](https://github.com/user-attachments/assets/5b1b7c2a-1368-4366-ba32-d59aeebf0d4d)

## SAM2

- Version: `large`
- Input type: `box input` 
- Multimask Output: `True`

![viz_7_sam2](https://github.com/user-attachments/assets/176124d8-2077-4b72-b87e-60da15614718)
![viz_13_sam2](https://github.com/user-attachments/assets/2f32e03a-ef01-426a-87d4-132d242685a1)

**Comment by heyoeyo** [comment]
It may be that the box is defined backwards, as in the top-left/bottom-right coordinates are reversed...? That might explain why the mask looks reversed.

From what I've seen, the results from v2 are generally similar to v1, but a bit more prone to weird artifacts. However, the new models scale to larger image sizes using a lot less VRAM than the v1 models, so they can give cleaner/smoother outlines.

**Comment by WaterKnight1998** [comment]
I am also seeing worse performance with points prediction

**Comment by heyoeyo** [comment]
> I am also seeing worse performance with points prediction

From what I've seen, between the different sized SAMv2 models, there can be significant differences in which masks (i.e. whole object, sub-components of object etc.) end up in the different indexes of the multi-mask output.
For example, the 0-th index mask of the large model tends to pick the smallest sub-component around the point prompt, while the same 0-th mask of the base-plus model tends to pick the 'whole' object. So you might be able to get a better result by picking a different mask output.

**Comment by rdfong** [comment]
![Screenshot from 2024-12-05 13-53-38](https://github.com/user-attachments/assets/32c5a71f-d0f1-406b-81d6-6e8216d0609e)

It also tends to output these strange artifact like results as seen here which was not an issue with sam v1, this is using a point prompt

**Comment by heyoeyo** [comment]
It could be that there's an issue with the coordinate position mismatching the image size, given that the mask seems to be selecting a different area than the point (or maybe the point is drawn in the wrong spot?).

If that point is correct, you can probably get a better result by using one of the other mask outputs. For example, if you're using the large model, that ground patch is cleanly segmented in the last-most mask (using multi-mask output):
![ground_patch_example](https://github.com/user-attachments/assets/9b386dc8-f4b5-4eea-b8b7-8b346434e7c8)
The non-multi-mask + 3 multi-mask outputs are shown on the right. You can see in one of the masks (second-last of multi-mask) it gives a patchy mask a bit like the example you showed, so switching to a different one may fix the problem.

The last-most mask isn't always the best though, for example the full car would be segmented by the second-last mask (and the last-most mask is patchy):
![car_example](https://github.com/user-attachments/assets/87a0e977-93e6-4b45-8c5e-4f9e7cc3d9c8)

Which mask is best also depends on the model size that's being used (the examples above are only for the v2.1 large model, the results can be very different for other model sizes).

**Comment by ludekcizinsky** [comment]
I experimented with the `Sam2ImagePredictor` and its hyperparameters, and found the following setup to work best:

1. Enable **multi-mask output** and select the mask with the highest score.  
2. Pass the selected mask through SAM for **at least two additional refinement steps**, each time feeding in the previous mask logits and again using multi-mask output to pick the best

This helped me to avoid missing parts of the mask or just weird irregularities in the mask. Hope this helps.
